### PR TITLE
 feat: expand IFNULL to support all types

### PR DESCRIPTION
**Pull request by big-andy-coates**
### Description 

Note: stacked on top of https://github.com/confluentinc/ksql/pull/4829: review that first, or just review the final commit in this PR: https://github.com/confluentinc/ksql/pull/4850/commits/2d2167d86c1f4fcb269f258b98c0c7dc9351bb08

fixes: #4831

`IFNULL` now supports all SQL types.

### Testing done 

Usual

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: docs-md/developer-guide/ksqldb-reference/scalar-functions.md [hunk 1]
```diff
@@ -431,6 +424,26 @@ UCASE
...
+
+Where the parameter type is a complex type, for example `ARRAY` or `STRUCT`, the contents of the
+complex type are not inspected. The behaviour is the same: the first non-null element is returned.
+
+IFNULL
+------
+
+`IFNULL(expression, altValue)`
+
+If the provided `expression` is NULL, returns `altValue`, otherwise, returns `expression`.
```
Comment (JimGalasyn) [inline docs-md/developer-guide/ksqldb-reference/scalar-functions.md]: Maybe `NULL` or `null`?

File: docs-md/developer-guide/ksqldb-reference/scalar-functions.md [hunk 2]
```diff
@@ -431,6 +424,26 @@ UCASE
...
 
+Nulls
+=====
+
+COALESCE
+--------
+
+`COALESCE(a, b, c, d)`
+
+Returns the first non-null parameter. All parameters must be of the same type.
```
Comment (JimGalasyn) [inline docs-md/developer-guide/ksqldb-reference/scalar-functions.md]: null or NULL? I'm fine with either, but we should be consistent.

**Comment by big-andy-coates** [comment]
cc @blueedgenick

**Comment by big-andy-coates** [comment]
> awesome - thanks for fixing it so quick @big-andy-coates !
> One minor comment, on which i won't block anything: having one UDF that basically delegates it's entire execution to another is obviously fair game from the standpoint that it gives correct output but it makes me a little uneasy. This is something we'd traditionally implement via a rewrite rule - so you literally wouldn't need an implementation class for IfNull at all. The end result is the same but it just "feels wrong" to me. Just saying.... :) Thinking more, I suppose we need the UDF impl class to have the function show up in the output of `list functions;` and to get the docs for it, which is an unfortunate consequence of how the whole UDF framework currently commingles all these capabilities. Hindsight is 20:20...

Exactly: its the `DESCRIBE function` et al _and_ the fact there will already be queries out there that utilize `IFNULL` that means we need to keep it around.

**Pull request merged**